tsl/src/remote/dist_copy.c

```diff
@@ -1248,355 +1225,369 @@ calculate_hyperspace_point_from_binary(Datum *values, bool *nulls, const Hypersp
-				data_node_rows = palloc(sizeof(DataNodeRows));
-				data_node_rows->connection = NULL;
-				data_node_rows->server_oid = chunk_data_node->foreign_server_oid;
-				data_node_rows->rows_total = 0;
-				/*
-				 * Not every tuple in a batch might be sent to every data node,
-				 * but we allocate the maximum possible size to avoid resizing.
-				 */
-				data_node_rows->row_indices = palloc(sizeof(int) * context->batch_row_count);
-				data_nodes = lappend(data_nodes, data_node_rows);
-			}
-
-			Assert(data_node_rows->server_oid == chunk_data_node->foreign_server_oid);
-
-			data_node_rows->row_indices[data_node_rows->rows_total] = row_in_batch;
-			data_node_rows->rows_total++;
-		}
-	}
+				else if (occurred[i].events & WL_SOCKET_WRITEABLE)
+				{
+					DataNodeConnection *dnc = occurred[i].user_data;
+					PGconn *pg_conn = remote_connection_get_pg_conn(dnc->connection);
 
-	/*
-	 * Flush the previous batch to avoid growing the outgoing buffers
-	 * indefinitely if some data node is not keeping up. It would be more
-	 * efficient to check for buffer growth and only flush then, but libpq
-	 * doesn't provide a way to know the outgoing buffer size. It also doesn't
-	 * provide any way to control the outgoing buffer size.
-	 * Don't do it if we have ended the COPY above to create new chunks.
-	 * The number 11 is an arbitrary prime, growing the output buffer to at
-	 * most 11ki rows sounds reasonable.
-	 */
-	if (!did_end_copy && context->batch_ordinal % 11 == 0)
-	{
-		flush_active_connections(&context->connection_state);
-	}
+					/* Flush and try sending again */
+					int ret = PQflush(pg_conn);
 
-	/*
-	 * Actually send the data to the data nodes. We don't interleave the data
-	 * nodes here, because the batches are relatively small.
-	 */
-	StringInfoData copy_data = { .data = palloc(MAX_BATCH_BYTES), .maxlen = MAX_BATCH_BYTES };
-	ListCell *lc;
-	foreach (lc, data_nodes)
-	{
-		DataNodeRows *dn = lfirst(lc);
-		TSConnectionId required_id = remote_connection_id(dn->server_oid, GetUserId());
-		Assert(dn->connection == NULL);
-		dn->connection = get_copy_connection_to_data_node(context, required_id);
-		PGconn *pg_conn = remote_connection_get_pg_conn(dn->connection);
-
-		resetStringInfo(&copy_data);
-		for (int row = 0; row < dn->rows_total; row++)
-		{
-			StringInfo row_data = context->batch_row_data[dn->row_indices[row]];
-			appendBinaryStringInfo(&copy_data, row_data->data, row_data->len);
-		}
-
-		/*
-		 * Send the copy data to the remote server.
-		 * It can't really return 0 ("would block") until it runs out
-		 * of memory. It just grows the buffer and tries to flush in
-		 * pqPutMsgEnd().
-		 */
-		int res = PQputCopyData(pg_conn, copy_data.data, copy_data.len);
+					if (ret == -1)
+						elog(ERROR,
+							 "error when flushing connection to data node %s: %s",
+							 remote_connection_node_name(dnc->connection),
+							 PQerrorMessage(pg_conn));
+				}
+			}
```

Comment:
Probably should be `ereport`. The `elog` is for internal program errors, and this looks like a network error.